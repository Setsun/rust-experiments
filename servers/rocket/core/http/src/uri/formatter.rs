use std::fmt;

use smallvec::SmallVec;

use uri::UriDisplay;

/// A struct used to format strings for [`UriDisplay`].
///
/// A mutable version of this struct is passed to [`UriDisplay::fmt()`]. This
/// struct properly formats series of named values for use in URIs. In
/// particular, this struct applies the following transformations:
///
///   * When **mutliple values** are written, they are separated by `&`.
///
///   * When a **named value** is written with [`write_named_value()`], the name
///     is written out, followed by a `=`, followed by the value.
///
///   * When **nested named values** are written, typically by passing a value
///     to [`write_named_value()`] whose implementation of `UriDisplay` also
///     calls `write_named_vlaue()`, the nested names are joined by a `.`,
///     written out followed by a `=`, followed by the value.
///
/// [`UriDisplay`]: uri::UriDisplay
/// [`UriDisplay::fmt()`]: uri::UriDisplay::fmt()
/// [`write_named_value()`]: uri::Formatter::write_named_value()
///
/// # Usage
///
/// Usage is fairly straightforward:
///
///   * For every _named value_ you wish to emit, call [`write_named_value()`].
///   * For every _unnamed value_ you wish to emit, call [`write_value()`].
///   * To write a string directly, call [`write_raw()`].
///
/// The `write_named_value` method automatically prefixes the `name` to the
/// written value and, along with `write_value` and `write_raw`, handles nested
/// calls to `write_named_value` automatically, prefixing names when necessary.
/// Unlike the other methods, `write_raw` does _not_ prefix any nested names
/// every time it is called. Instead, it only prefixes names the _first_ time it
/// is called, after a call to `write_named_value` or `write_value`, or after a
/// call to [`refresh()`].
///
/// [`refresh()`]: uri::Formatter::refresh()
///
/// # Example
///
/// The following example uses all of the `write` methods in a varied order to
/// display the semantics of `Formatter`. Note that `UriDisplay` should rarely
/// be implemented manually, preferring to use the derive, and that this
/// implementation is purely demonstrative.
///
/// ```rust
/// # extern crate rocket;
/// use std::fmt;
///
/// use rocket::http::uri::{Formatter, UriDisplay};
///
/// struct Outer {
///     value: Inner,
///     another: usize,
///     extra: usize
/// }
///
/// struct Inner {
///     value: usize,
///     extra: usize
/// }
///
/// impl UriDisplay for Outer {
///     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
///         f.write_named_value("outer_field", &self.value)?;
///         f.write_named_value("another", &self.another)?;
///         f.write_raw("out")?;
///         f.write_raw("side")?;
///         f.write_value(&self.extra)
///     }
/// }
///
/// impl UriDisplay for Inner {
///     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
///         f.write_named_value("inner_field", &self.value)?;
///         f.write_value(&self.extra)?;
///         f.write_raw("inside")
///     }
/// }
///
/// let inner = Inner { value: 0, extra: 1 };
/// let outer = Outer { value: inner, another: 2, extra: 3 };
/// let uri_string = format!("{}", &outer as &UriDisplay);
/// assert_eq!(uri_string, "outer_field.inner_field=0&\
///                         outer_field=1&\
///                         outer_field=inside&\
///                         another=2&\
///                         outside&\
///                         3");
/// ```
///
/// Note that you can also use the `write!` macro to write directly to the
/// formatter as long as the [`std::fmt::Write`] trait is in scope. Internally,
/// the `write!` macro calls [`write_raw()`], so care must be taken to ensure
/// that the written string is URI-safe.
///
/// ```rust
/// # #[macro_use] extern crate rocket;
/// use std::fmt::{self, Write};
///
/// use rocket::http::uri::{Formatter, UriDisplay};
///
/// pub struct Complex(u8, u8);
///
/// impl UriDisplay for Complex {
///     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
///         write!(f, "{}+{}", self.0, self.1)
///     }
/// }
///
/// #[derive(UriDisplay)]
/// struct Message {
///     number: Complex,
/// }
///
/// let message = Message { number: Complex(42, 231) };
/// let uri_string = format!("{}", &message as &UriDisplay);
/// assert_eq!(uri_string, "number=42+231");
/// ```
///
/// [`write_value()`]: uri::Formatter::write_value()
/// [`write_raw()`]: uri::Formatter::write_raw()
pub struct Formatter<'i, 'f: 'i> {
    crate prefixes: SmallVec<[&'static str; 3]>,
    crate inner: &'i mut fmt::Formatter<'f>,
    crate previous: bool,
    crate fresh: bool
}

impl<'i, 'f: 'i> Formatter<'i, 'f> {
    crate fn new(formatter: &'i mut fmt::Formatter<'f>) -> Self {
        Formatter {
            prefixes: SmallVec::new(),
            inner: formatter,
            previous: false,
            fresh: true,
        }
    }

    fn with_prefix<F>(&mut self, prefix: &str, f: F) -> fmt::Result
        where F: FnOnce(&mut Self) -> fmt::Result
    {
        // TODO: PROOF OF CORRECTNESS.
        let prefix: &'static str = unsafe { ::std::mem::transmute(prefix) };

        self.prefixes.push(prefix);
        let result = f(self);
        self.prefixes.pop();

        result
    }

    #[inline(always)]
    fn refreshed<F: FnOnce(&mut Self) -> fmt::Result>(&mut self, f: F) -> fmt::Result {
        self.refresh();
        let result = f(self);
        self.refresh();
        result
    }

    /// Writes `string` to `self`.
    ///
    /// If `self` is _fresh_ (after a call to other `write_` methods or
    /// [`refresh()`]), prefixes any names as necessary.
    ///
    /// This method is called by the `write!` macro.
    ///
    /// [`refresh()`]: Formatter::refresh()
    ///
    /// # Example
    ///
    /// ```rust
    /// # extern crate rocket;
    /// use std::fmt;
    ///
    /// use rocket::http::uri::{Formatter, UriDisplay};
    ///
    /// struct Foo;
    ///
    /// impl UriDisplay for Foo {
    ///     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    ///         f.write_raw("f")?;
    ///         f.write_raw("o")?;
    ///         f.write_raw("o")
    ///     }
    /// }
    ///
    /// let foo = Foo;
    /// let uri_string = format!("{}", &foo as &UriDisplay);
    /// assert_eq!(uri_string, "foo");
    /// ```
    pub fn write_raw<S: AsRef<str>>(&mut self, string: S) -> fmt::Result {
        let s = string.as_ref();
        if self.fresh {
            if self.previous {
                self.inner.write_str("&")?;
            }

            if !self.prefixes.is_empty() {
                for (i, prefix) in self.prefixes.iter().enumerate() {
                    self.inner.write_str(prefix)?;
                    if i < self.prefixes.len() - 1 {
                        self.inner.write_str(".")?;
                    }
                }

                self.inner.write_str("=")?;
            }
        }

        self.fresh = false;
        self.previous = true;
        self.inner.write_str(s)
    }

    /// Writes the named value `value` by prefixing `name` followed by `=` to
    /// the value. Any nested names are also prefixed as necessary.
    ///
    /// Refreshes `self` before the name is written and after the value is
    /// written.
    ///
    /// # Example
    ///
    /// ```rust
    /// # extern crate rocket;
    /// use std::fmt;
    ///
    /// use rocket::http::uri::{Formatter, UriDisplay};
    ///
    /// struct Foo {
    ///     name: usize
    /// }
    ///
    /// impl UriDisplay for Foo {
    ///     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    ///         f.write_named_value("name", &self.name)
    ///     }
    /// }
    ///
    /// let foo = Foo { name: 123 };
    /// let uri_string = format!("{}", &foo as &UriDisplay);
    /// assert_eq!(uri_string, "name=123");
    /// ```
    #[inline]
    pub fn write_named_value<T: UriDisplay>(&mut self, name: &str, value: T) -> fmt::Result {
        self.refreshed(|f| f.with_prefix(name, |f| f.write_value(value)))
    }

    /// Writes the unnamed value `value`. Any nested names are prefixed as
    /// necessary.
    ///
    /// Refreshes `self` before and after the value is written.
    ///
    /// # Example
    ///
    /// ```rust
    /// # extern crate rocket;
    /// use std::fmt;
    ///
    /// use rocket::http::uri::{Formatter, UriDisplay};
    ///
    /// struct Foo(usize);
    ///
    /// impl UriDisplay for Foo {
    ///     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    ///         f.write_value(&self.0)
    ///     }
    /// }
    ///
    /// let foo = Foo(123);
    /// let uri_string = format!("{}", &foo as &UriDisplay);
    /// assert_eq!(uri_string, "123");
    /// ```
    #[inline]
    pub fn write_value<T: UriDisplay>(&mut self, value: T) -> fmt::Result {
        self.refreshed(|f| UriDisplay::fmt(&value, f))
    }

    /// Refreshes the formatter.
    ///
    /// After refreshing, [`write_raw()`] will prefix any nested names as well
    /// as insert an `&` separator.
    ///
    /// [`write_raw()`]: Formatter::write_raw()
    ///
    /// # Example
    ///
    /// ```rust
    /// # #[macro_use] extern crate rocket;
    /// use std::fmt;
    ///
    /// use rocket::http::uri::{Formatter, UriDisplay};
    ///
    /// struct Foo;
    ///
    /// impl UriDisplay for Foo {
    ///     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    ///         f.write_raw("a")?;
    ///         f.write_raw("raw")?;
    ///         f.refresh();
    ///         f.write_raw("format")
    ///     }
    /// }
    ///
    /// #[derive(UriDisplay)]
    /// struct Message {
    ///     inner: Foo,
    /// }
    ///
    /// let msg = Message { inner: Foo };
    /// let uri_string = format!("{}", &msg as &UriDisplay);
    /// assert_eq!(uri_string, "inner=araw&inner=format");
    /// ```
    #[inline(always)]
    pub fn refresh(&mut self) {
        self.fresh = true;
    }
}

impl<'f, 'i: 'f> fmt::Write for Formatter<'f, 'i> {
    fn write_str(&mut self, s: &str) -> fmt::Result {
        self.write_raw(s)
    }
}
