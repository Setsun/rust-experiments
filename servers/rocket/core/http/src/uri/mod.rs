//! Types for URIs and traits for rendering URI components.

mod uri;
mod uri_display;
mod formatter;
mod from_uri_param;
mod origin;
mod authority;
mod absolute;
mod segments;

pub use parse::uri::Error;

pub use self::uri::*;
pub use self::authority::*;
pub use self::origin::*;
pub use self::absolute::*;
pub use self::uri_display::*;
pub use self::formatter::*;
pub use self::from_uri_param::*;
pub use self::segments::*;
