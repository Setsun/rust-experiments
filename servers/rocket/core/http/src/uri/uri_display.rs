use std::fmt;
use std::path::{Path, PathBuf};
use std::borrow::Cow;

use percent_encoding::{utf8_percent_encode, DEFAULT_ENCODE_SET};

use {RawStr, uri::{Uri, Formatter}, ext::Normalize};

mod priv_encode_set {
    /// This encode set is used for strings where '/' characters are known to be
    /// safe; all other special path segment characters are encoded.
    define_encode_set! { pub PATH_ENCODE_SET = [super::DEFAULT_ENCODE_SET] | {'%'} }
}

use self::priv_encode_set::PATH_ENCODE_SET;

/// Trait implemented by types that can be displayed as part of a URI in `uri!`.
///
/// Types implementing this trait can be displayed in a URI-safe manner. Unlike
/// `Display`, the string written by a `UriDisplay` implementation must be
/// URI-safe. In practice, this means that the string must either be
/// percent-encoded or consist only of characters that are alphanumeric, "-",
/// ".", "_", or "~" - the "unreserved" characters.
///
/// # Code Generation
///
/// When the `uri!` macro is used to generate a URI for a route, the types for
/// the route's URI parameters must implement `UriDisplay`. The `UriDisplay`
/// implementation for these types is used when generating the URI.
///
/// To illustrate `UriDisplay`'s role in code generation for `uri!`, consider
/// the following route:
///
/// ```rust
/// # #![feature(proc_macro_hygiene, decl_macro)]
/// # #[macro_use] extern crate rocket;
/// #[get("/item/<id>?<track>")]
/// fn get_item(id: i32, track: String) { /* .. */ }
/// ```
///
/// A URI for this route can be generated as follows:
///
/// ```rust
/// # #![feature(proc_macro_hygiene, decl_macro)]
/// # #[macro_use] extern crate rocket;
/// # type T = ();
/// # #[get("/item/<id>?<track>")]
/// # fn get_item(id: i32, track: String) { /* .. */ }
/// #
/// // With unnamed parameters.
/// uri!(get_item: 100, "inbound");
///
/// // With named parameters.
/// uri!(get_item: id = 100, track = "inbound");
/// uri!(get_item: track = "inbound", id = 100);
/// ```
///
/// After verifying parameters and their types, Rocket will generate code
/// similar to the following:
///
/// ```rust
/// # extern crate rocket;
/// # use rocket::http::uri::UriDisplay;
/// #
/// format!("/item/{}?track={}", &100 as &UriDisplay, &"inbound" as &UriDisplay);
/// ```
///
/// For this expression to typecheck, both `i32` and `Value` must implement
/// `UriDisplay`. As can be seen, the implementation will be used to display the
/// value in a URI-safe manner.
///
/// [`uri!`]: /rocket_codegen/#typed-uris-uri
///
/// # Provided Implementations
///
/// Rocket implements `UriDisplay` for several built-in types. Their behavior is
/// documented here.
///
///   * **i8, i16, i32, i64, isize, u8, u16, u32, u64, usize, f32, f64, bool,
///     IpAddr, Ipv4Addr, Ipv6Addr**
///
///     The implementation of `UriDisplay` for these types is identical to the
///     `Display` implementation.
///
///   * **[`&RawStr`](RawStr), `String`, `&str`, `Cow<str>`**
///
///     The string is percent encoded.
///
///   * **`&T`, `&mut T`** _where_ **`T: UriDisplay`**
///
///     Uses the implementation of `UriDisplay` for `T`.
///
/// # Deriving
///
/// Manually implementing `UriDisplay` should be done with care. For most use
/// cases, deriving `UriDisplay` will suffice:
///
/// ```rust
/// # #[macro_use] extern crate rocket;
/// # use rocket::http::uri::UriDisplay;
/// #[derive(FromForm, UriDisplay)]
/// struct User {
///     name: String,
///     age: usize,
/// }
///
/// let user = User { name: "Michael Smith".into(), age: 31 };
/// let uri_string = format!("{}", &user as &UriDisplay);
/// assert_eq!(uri_string, "name=Michael%20Smith&age=31");
/// ```
///
/// As long as every field in the structure (or enum) implements `UriDisplay`,
/// the trait can be derived. The implementation calls
/// [`Formatter::write_named_value()`] for every named field and
/// [`Formatter::write_value()`] for every unnamed field. See the [`UriDisplay`
/// derive] documentation for full details.
///
/// [`UriDisplay` derive]: ../../../rocket_codegen/derive.UriDisplay.html
///
/// # Implementing
///
/// Implementing `UriDisplay` is similar to implementing
/// [`Display`](::std::fmt::Display) with the caveat that extra care must be
/// taken to ensure that the written string is URI-safe. As mentioned before, in
/// practice, this means that the string must either be percent-encoded or
/// consist only of characters that are alphanumeric, "-", ".", "_", or "~".
///
/// When manually implementing `UriDisplay` for your types, you should defer to
/// existing implementations of `UriDisplay` as much as possible. In the example
/// below, for instance, `Name`'s implementation defers to `String`'s
/// implementation. To percent-encode a string, use
/// [`Uri::percent_encode()`](uri::Uri::percent_encode()).
///
/// ## Example
///
/// The following snippet consists of a `Name` type that implements both
/// `FromParam` and `UriDisplay`. The `FromParam` implementation allows `Name`
/// to be used as the target type of a dynamic parameter, while the `UriDisplay`
/// implementation allows URIs to be generated for routes with `Name` as a
/// dynamic parameter type. Note the custom parsing in the `FromParam`
/// implementation; as a result of this, a custom (reflexive) `UriDisplay`
/// implementation is required.
///
/// ```rust
/// # #![feature(proc_macro_hygiene, decl_macro)]
/// # #[macro_use] extern crate rocket;
/// use rocket::http::RawStr;
/// use rocket::request::FromParam;
///
/// struct Name(String);
///
/// const PREFIX: &str = "name:";
///
/// impl<'r> FromParam<'r> for Name {
///     type Error = &'r RawStr;
///
///     /// Validates parameters that start with 'name:', extracting the text
///     /// after 'name:' as long as there is at least one character.
///     fn from_param(param: &'r RawStr) -> Result<Self, Self::Error> {
///         let decoded = param.percent_decode().map_err(|_| param)?;
///         if !decoded.starts_with(PREFIX) || decoded.len() < (PREFIX.len() + 1) {
///             return Err(param);
///         }
///
///         let real_name = decoded[PREFIX.len()..].to_string();
///         Ok(Name(real_name))
///     }
/// }
///
/// use std::fmt;
/// use rocket::http::uri::{Formatter, UriDisplay};
/// use rocket::response::Redirect;
///
/// impl UriDisplay for Name {
///     /// Delegates to the `UriDisplay` implementation for `String` to ensure
///     /// that the written string is URI-safe. In this case, the string will
///     /// be percent encoded. Prefixes the inner name with `name:`.
///     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
///         f.write_value(&format!("name:{}", self.0))
///     }
/// }
///
/// #[get("/name/<name>")]
/// fn redirector(name: Name) -> Redirect {
///     Redirect::to(uri!(real: name))
/// }
///
/// #[get("/<name>")]
/// fn real(name: Name) -> String {
///     format!("Hello, {}!", name.0)
/// }
///
/// let uri = uri!(real: Name("Mike Smith".into()));
/// assert_eq!(uri.path(), "/name:Mike%20Smith");
/// ```
pub trait UriDisplay {
    /// Formats `self` in a URI-safe manner using the given formatter.
    fn fmt(&self, f: &mut Formatter) -> fmt::Result;
}

impl<'a> fmt::Display for &'a UriDisplay {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        UriDisplay::fmt(*self, &mut Formatter::new(f))
    }
}

/// Percent-encodes the raw string.
impl UriDisplay for RawStr {
    #[inline(always)]
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        f.write_raw(&Uri::percent_encode(self.as_str()))
    }
}

/// Percent-encodes the raw string.
impl UriDisplay for str {
    #[inline(always)]
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        f.write_raw(&Uri::percent_encode(self))
    }
}

/// Percent-encodes the raw string.
impl<'a> UriDisplay for Cow<'a, str> {
    #[inline(always)]
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        f.write_raw(&Uri::percent_encode(self))
    }
}

/// Percent-encodes the raw string.
impl UriDisplay for String {
    #[inline(always)]
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        f.write_raw(&Uri::percent_encode(self.as_str()))
    }
}

/// Percent-encodes each segment in the path and normalizes separators.
impl UriDisplay for PathBuf {
    #[inline]
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        let string = self.normalized_str();
        let enc: Cow<str> = utf8_percent_encode(&string, PATH_ENCODE_SET).into();
        f.write_raw(&enc)
    }
}

/// Percent-encodes each segment in the path and normalizes separators.
impl UriDisplay for Path {
    #[inline]
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        let string = self.normalized_str();
        let enc: Cow<str> = utf8_percent_encode(&string, PATH_ENCODE_SET).into();
        f.write_raw(&enc)
    }
}

macro_rules! impl_with_display {
    ($($T:ty),+) => {$(
        /// This implementation is identical to the `Display` implementation.
        impl UriDisplay for $T  {
            #[inline(always)]
            fn fmt(&self, f: &mut Formatter) -> fmt::Result {
                use std::fmt::Write;
                write!(f, "{}", self)
            }
        }
    )+}
}

use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};

impl_with_display! {
    i8, i16, i32, i64, i128, isize,
    u8, u16, u32, u64, u128, usize,
    f32, f64, bool,
    IpAddr, Ipv4Addr, Ipv6Addr
}

macro_rules! impl_for_ref {
    ($($T:ty),+) => {$(
        /// Uses the implementation of `UriDisplay` for `T`.
        impl<'a, T: UriDisplay + ?Sized> UriDisplay for $T {
            #[inline(always)]
            fn fmt(&self, f: &mut Formatter) -> fmt::Result {
                UriDisplay::fmt(*self, f)
            }
        }
    )+}
}

impl_for_ref!(&'a mut T, &'a T);
