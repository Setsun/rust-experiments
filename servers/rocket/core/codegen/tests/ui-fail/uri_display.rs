#[macro_use] extern crate rocket;

#[derive(UriDisplay)]
struct Foo1;
//~^ ERROR not supported

#[derive(UriDisplay)]
struct Foo2();
//~^ ERROR not supported

#[derive(UriDisplay)]
enum Foo3 { }
//~^ ERROR not supported

#[derive(UriDisplay)]
enum Foo4 {
    Variant,
    //~^ ERROR not supported
}

#[derive(UriDisplay)]
struct Foo5(String, String);
//~^ ERROR exactly one

#[derive(UriDisplay)]
struct Foo6 {
    #[form(field = 123)]
    //~^ ERROR invalid value: expected string
    field: String,
}
