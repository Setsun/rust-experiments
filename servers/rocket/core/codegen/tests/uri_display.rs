#![feature(proc_macro_hygiene, decl_macro)]

#[macro_use] extern crate rocket;

use rocket::http::RawStr;
use rocket::http::uri::UriDisplay;

macro_rules! assert_uri_display {
    ($v:expr, $s:expr) => (
        let uri_string = format!("{}", &$v as &UriDisplay);
        assert_eq!(uri_string, $s);
    )
}

#[derive(UriDisplay, Clone)]
enum Foo<'r> {
    First(&'r RawStr),
    Second {
        inner: &'r RawStr,
        other: usize,
    },
    Third {
        #[form(field = "type")]
        kind: String,
    },
}

#[test]
fn uri_display_foo() {
    let foo = Foo::First("hello".into());
    assert_uri_display!(foo, "hello");

    let foo = Foo::First("hello there".into());
    assert_uri_display!(foo, "hello%20there");

    let foo = Foo::Second { inner: "hi".into(), other: 123 };
    assert_uri_display!(foo, "inner=hi&other=123");

    let foo = Foo::Second { inner: "hi bo".into(), other: 321 };
    assert_uri_display!(foo, "inner=hi%20bo&other=321");

    let foo = Foo::Third { kind: "hello".into() };
    assert_uri_display!(foo, "type=hello");

    let foo = Foo::Third { kind: "hello there".into() };
    assert_uri_display!(foo, "type=hello%20there");
}

#[derive(UriDisplay)]
struct Bar<'a> {
    foo: Foo<'a>,
    baz: String,
}

#[test]
fn uri_display_bar() {
    let foo = Foo::First("hello".into());
    let bar = Bar { foo, baz: "well, hi!".into() };
    assert_uri_display!(bar, "foo=hello&baz=well,%20hi!");

    let foo = Foo::Second { inner: "hi".into(), other: 123 };
    let bar = Bar { foo, baz: "done".into() };
    assert_uri_display!(bar, "foo.inner=hi&foo.other=123&baz=done");

    let foo = Foo::Third { kind: "hello".into() };
    let bar = Bar { foo, baz: "turkey day".into() };
    assert_uri_display!(bar, "foo.type=hello&baz=turkey%20day");
}

#[derive(UriDisplay)]
struct Baz<'a> {
    foo: Foo<'a>,
    bar: Bar<'a>,
    last: String
}

#[test]
fn uri_display_baz() {
    let foo1 = Foo::Second { inner: "hi".into(), other: 123 };
    let foo2 = Foo::Second { inner: "bye".into(), other: 321 };
    let bar = Bar { foo: foo2, baz: "done".into() };
    let baz = Baz { foo: foo1, bar, last: "ok".into() };
    assert_uri_display!(baz, "foo.inner=hi&foo.other=123&\
                              bar.foo.inner=bye&bar.foo.other=321&bar.baz=done&\
                              last=ok");

    let foo1 = Foo::Third { kind: "hello".into() };
    let foo2 = Foo::First("bye".into());
    let bar = Bar { foo: foo1, baz: "end".into() };
    let baz = Baz { foo: foo2, bar, last: "done".into() };
    assert_uri_display!(baz, "foo=bye&\
                              bar.foo.type=hello&bar.baz=end&\
                              last=done");
}
