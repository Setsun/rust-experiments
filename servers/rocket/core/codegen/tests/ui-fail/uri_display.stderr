error: fieldless structs or variants are not supported
 --> $DIR/uri_display.rs:4:1
  |
4 | struct Foo1;
  | ^^^^^^^^^^^^
  |
note: error occurred while deriving `UriDisplay`
 --> $DIR/uri_display.rs:3:10
  |
3 | #[derive(UriDisplay)]
  |          ^^^^^^^^^^

error: fieldless structs or variants are not supported
 --> $DIR/uri_display.rs:8:1
  |
8 | struct Foo2();
  | ^^^^^^^^^^^^^^
  |
note: error occurred while deriving `UriDisplay`
 --> $DIR/uri_display.rs:7:10
  |
7 | #[derive(UriDisplay)]
  |          ^^^^^^^^^^

error: empty enums are not supported
  --> $DIR/uri_display.rs:12:1
   |
12 | enum Foo3 { }
   | ^^^^^^^^^^^^^
   |
note: error occurred while deriving `UriDisplay`
  --> $DIR/uri_display.rs:11:10
   |
11 | #[derive(UriDisplay)]
   |          ^^^^^^^^^^

error: fieldless structs or variants are not supported
  --> $DIR/uri_display.rs:17:5
   |
17 |     Variant,
   |     ^^^^^^^
   |
note: error occurred while deriving `UriDisplay`
  --> $DIR/uri_display.rs:15:10
   |
15 | #[derive(UriDisplay)]
   |          ^^^^^^^^^^

error: tuple structs or variants must have exactly one field
  --> $DIR/uri_display.rs:22:12
   |
22 | struct Foo5(String, String);
   |            ^^^^^^^^^^^^^^^^
   |
note: error occurred while deriving `UriDisplay`
  --> $DIR/uri_display.rs:21:10
   |
21 | #[derive(UriDisplay)]
   |          ^^^^^^^^^^

error: invalid value: expected string literal
  --> $DIR/uri_display.rs:27:20
   |
27 |     #[form(field = 123)]
   |                    ^^^
   |
note: error occurred while deriving `UriDisplay`
  --> $DIR/uri_display.rs:25:10
   |
25 | #[derive(UriDisplay)]
   |          ^^^^^^^^^^

error: aborting due to 6 previous errors

